neural_cherche/retrieve/bm25.py

```diff
@@ -0,0 +1,203 @@
+    ...     key="id",
+    ...     on=["document"],
+    ... )
+
+    >>> documents_embeddings = retriever.encode_documents(
+    ...     documents=documents
+    ... )
+
+    >>> retriever = retriever.add(
+    ...     documents_embeddings=documents_embeddings,
+    ... )
+
+    >>> queries_embeddings = retriever.encode_queries(
+    ...     queries=queries
+    ... )
+
+    >>> scores = retriever(
+    ...     queries_embeddings=queries_embeddings,
+    ...     k=4
+    ... )
+
+    >>> pprint(scores)
+    [[{'id': 0, 'similarity': 1.0}],
+     [{'id': 1, 'similarity': 0.9999999999999999}],
+     [{'id': 2, 'similarity': 0.9999999999999999}]]
+
+    References
+    ----------
+    1. [sklearn.feature_extraction.text.CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html)
+    2. [Python: tf-idf-cosine: to find document similarity](https://stackoverflow.com/questions/12118720/python-tf-idf-cosine-to-find-document-similarity)
+
+    """
+    def __init__(
+        self,
+        key: str,
+        on: list[str],
+        b=0.75, 
+        k1=1.5,
+        countVector=None,
+    ) -> None:
+        self.key = key
+        self.on = [on] if isinstance(on, str) else on
+        self.countVector=(
+            CountVectorizer(lowercase=True,ngram_range=(3, 7), analyzer="char")
+            if countVector is None
+            else countVector
+        )
+        #Tunable parameters
+        self.b = b # The impact of document length normalization, Higher --> penalize longer documents more.
+        self.k1 = k1 #how quickly the impact of term frequency saturates, Higher --> make term frequency more influential
+        self.matrix = None
+        self.documents = []
+        self.n_documents = 0
+
+    def encode_documents(self, documents: list[dict]) -> dict[str, csr_matrix]:
+        """Encode queries into sparse matrix.
+
+        Parameters
+        ----------
+        documents
+            Documents to encode.
+
+        """
+        content = [
+            " ".join([doc.get(field, "") for field in self.on]) for doc in documents
+        ]
+
+        self.matrix = self.countVector.fit_transform(raw_documents=content)
+        self.fit=True
+
+        return {document[self.key]: row for document, row in zip(documents, self.matrix)}
+    def add(
+        self,
+        documents_embeddings: dict[str, csr_matrix],
+    ) -> "CountVectorizer":
+        """Add new documents to the CountVectorizer retriever."""
+        for document_key in documents_embeddings:
+            self.documents.append({self.key: document_key})
+
+        self.n_documents += len(documents_embeddings)
+        n_samples,n_features  = self.matrix.shape
+        df = np.bincount(self.matrix.indices, minlength=n_features)#Count the number of non-zero values for each feature in sparse X
+        idf = np.log(n_samples / (df)) #compute the log of idf
+        avdl  = self.matrix.sum(1).mean() # mean of all 
```

Comment:
Would it make sense to call self.matrix.sum(1) once and then call .mean() and .A1 to get avdl and len_X ? If self.matrix.sum(1) does not cost anything to compute we can keep it like this